### PR TITLE
read only field support

### PR DESCRIPTION
**Pull request by rolandknight**
Support for read only fields. Fields with read only annotation (requires this pending contrib MR: https://github.com/ent/contrib/pull/260) will have these parameters suppressed (useful for createdBy and createdAt fields).

Example schema:

// Fields of the BaseMixin.
func (BaseMixin) Fields() []ent.Field {
  return []ent.Field{
    field.Time("created_at").Default(time.Now).Immutable().Annotations(entoas.Annotation{Field: entoas.FieldConfig{ReadOnly: true}}),
    field.Time("updated_at").Default(time.Now).Annotations(entoas.Annotation{Field: entoas.FieldConfig{ReadOnly: true}}),
  }
}

### CODE REVIEW

File: template/create.tmpl [hunk 1]
```diff
@@ -2,16 +2,19 @@
...
-				{{- $x := "v" }}
-				{{- if $f.IsEnum }}{{ $x = convertTo $f.Type.String $x }}{{ end }}
-				b.Set{{ $f.StructField }}({{ $x }})
-			}
-		{{- else }}
-			{{- $x := print "req." $f.StructField }}
-			{{- if $f.IsEnum }}{{ $x = convertTo $f.Type.String $x }}{{ end }}
-			b.Set{{ $f.StructField }}({{ $x }})
+	    {{- $a := fieldAnnotation $f }}
+	    {{- if or (not $a) (not $a.Field.ReadOnly) }}
```
Comment (masseelch) [inline template/create.tmpl:6]: Please apply De'Morgan here :wink:

File: template/update.tmpl [hunk 2]
```diff
@@ -4,7 +4,8 @@
 	b := h.client.{{ $.Name }}.UpdateOneID(params.{{ $.ID.StructField }})
 	// Add all fields.
 	{{- range $f := $.Fields }}
-		{{- if not $f.Immutable }}
+	    {{- $a := fieldAnnotation $f }}
+	    {{- if and (not $f.Immutable) (or (not $a) (not $a.Field.ReadOnly)) }}
```
Comment (masseelch) [inline template/update.tmpl:8]: Here again.

**Review by masseelch — CHANGES_REQUESTED** [review]
Hey @rolandknight, I added two comments already. But lets first finish our work in `ent/contrib` before we continue this :+1:

File: template/create.tmpl (hunk 1 above)
Comment (rolandknight) [reply template/create.tmpl]: yes. was lazy there. DeMorgan's applied. should be ready to merge now

File: template/update.tmpl (hunk 2 above)
Comment (rolandknight) [reply template/update.tmpl]: fixed

File: template.go [hunk 3]
```diff
@@ -19,25 +19,26 @@ var (
...
+		"isUpdate":        isUpdate,
+		"isDelete":        isDelete,
+		"isList":          isList,
+		"kebab":           strcase.KebabCase,
+		"nodeOperations":  entoas.NodeOperations,
+		"replaceAll":      strings.ReplaceAll,
+		"setFieldExpr":    setFieldExpr,
+		"viewName":        entoas.ViewName,
+		"viewNameEdge":    entoas.ViewNameEdge,
+		"fieldAnnotation": entoas.FieldAnnotation,
```
Comment (masseelch) [inline template.go:41]: Please order alphabetically

File: template/create.tmpl [hunk 4]
```diff
@@ -2,16 +2,19 @@
...
+	    		if v, ok := req.{{ $f.StructField}}.Get(); ok {
+		    		{{- $x := "v" }}
+			    	{{- if $f.IsEnum }}{{ $x = convertTo $f.Type.String $x }}{{ end }}
+				    b.Set{{ $f.StructField }}({{ $x }})
+			    }
+		    {{- else }}
+			    {{- $x := print "req." $f.StructField }}
+			    {{- if $f.IsEnum }}{{ $x = convertTo $f.Type.String $x }}{{ end }}
+			    b.Set{{ $f.StructField }}({{ $x }})
+		    {{- end }}
```
Comment (masseelch) [inline template/create.tmpl:17]: fix indention

File: template/create.tmpl [hunk 5]
```diff
@@ -2,16 +2,19 @@
...
-				{{- $x := "v" }}
-				{{- if $f.IsEnum }}{{ $x = convertTo $f.Type.String $x }}{{ end }}
-				b.Set{{ $f.StructField }}({{ $x }})
-			}
-		{{- else }}
-			{{- $x := print "req." $f.StructField }}
-			{{- if $f.IsEnum }}{{ $x = convertTo $f.Type.String $x }}{{ end }}
-			b.Set{{ $f.StructField }}({{ $x }})
+	    {{- $a := fieldAnnotation $f }}
+	    {{- if not (and $a $a.ReadOnly) }}
```
Comment (masseelch) [inline template/create.tmpl:6]: DeMorgan -> makes it more readable in this case :-)

File: template/update.tmpl [hunk 6]
```diff
@@ -4,7 +4,8 @@
 	b := h.client.{{ $.Name }}.UpdateOneID(params.{{ $.ID.StructField }})
 	// Add all fields.
 	{{- range $f := $.Fields }}
-		{{- if not $f.Immutable }}
+	    {{- $a := fieldAnnotation $f }}
+	    {{- if not (or $f.Immutable (and $a $a.ReadOnly)) }}
```
Comment (masseelch) [inline template/update.tmpl:8]: DeMorgan again :-)

**Review by masseelch — CHANGES_REQUESTED** [review]
Nice! Can you add a test in `internal/integration/ogent`?

File: template.go (hunk 3 above)
Comment (rolandknight) [reply template.go]: done. a few other were out of order as well

File: template/create.tmpl (hunk 5 above)
Comment (rolandknight) [reply template/create.tmpl]: agreed and done

**Comment by rolandknight** [comment]
> Nice! Can you add a test in `internal/integration/ogent`?

FYI, as a test category.go has a read only field: https://github.com/ariga/ogent/pull/65/files#diff-bcbd7b4be1e34f2b3cffc7263df9a0ce54ff9830760d871a7c3ef8e7dff5ff9b

**Comment by masseelch** [comment]
Very nice work! Thank you @rolandknight :blush:

**Pull request merged**